### PR TITLE
Commit various enhancements

### PR DESCRIPTION
**Pull request by dcalve**
I've been making a few changes to the cdftools as required for an update to ocean validation software. In summary, these are:

#### Enhancements
* Write time coordinate variables as double precision
  * Several supporting functions have been added in `cdfio`
  * This has **only been applied to `cdfmean` and `cdfmoc`** at the moment
* Added basin mask support and output file specification to `cdfmean`
  * New syntax: `cdfmean ... -out_suff output_file_suffix -basins n_basins basin_names`

#### Bug fixes
* Fix bug with `getvar` in `cdfio` when called for mesh mask files
  * The `getvar` function calls `getdim(...,cn_x)`, which does not work when accessing mesh mask variables
  * Since mesh mask dimension names are still hard-coded, I have added a get_dim(..., 'x') if the above fails
* Fix `cn_v...msk` namelist variables not being referenced in `cdfmoc`
* Add missing namelists and namelist variables in `modcdfnames`

Would these have value to the cdftools community? 

cc @TimGraham1982 for interest

### CODE REVIEW

**Comment by pmathiot** [comment]
Yes I think this could be useful as long as this is not mandatory. Some other routine could be concerned by your change but you don't need to do it now, we will do it when we need it.

**Review by pmathiot — COMMENTED** [review]
I am wondering why we need to change the time variable routine (getvar1d) from REAL(4) to REAL(8).

**Comment by dcalve** [comment]
There is truncation of double precision variables read from a file if `REAL(4)` is used in `getvar1d`. XIOS always outputs time coordinates in double precision. It is optional behaviour in the function, however.

File: src/modcdfnames.f90 [hunk 1]
```diff
@@ -276,6 +276,7 @@ SUBROUTINE ReadCdfNames ()
     READ(inam, namvars    )
     READ(inam, nambathy   )
     READ(inam, namsqdvar  )
+    READ(inam, namcubvar  )
```
Comment (pmathiot) [inline src/modcdfnames.f90:279]: Is this something needed for your change or a bug you found when you worked on it ?

Comment (dcalve) [reply src/modcdfnames.f90:279]: That's something I noticed while looking at the code

File: src/modcdfnames.f90 [hunk 2]
```diff
@@ -190,8 +190,8 @@ MODULE modCdfNames
     NAMELIST/namvars/ cn_sohefldo, cn_solhflup, cn_sosbhfup
     NAMELIST/namvars/ cn_solwfldo, cn_soshfldo
     NAMELIST/namvars/ cn_sowaflup, cn_sowaflcd, cn_sowafldp, cn_iowaflup
-    NAMELIST/namvars/ cn_zomsfatl, cn_zomsfglo, cn_zomsfpac, cn_zomsfinp, cn_zomsfind
-    NAMELIST/namvars/ cn_zoisoatl, cn_zoisoglo, cn_zoisopac, cn_zoisoinp, cn_zoisoind
+    NAMELIST/namvars/ cn_zomsfatl, cn_zomsfglo, cn_zomsfpac, cn_zomsfinp, cn_zomsfind, cn_zomsfinp0
+    NAMELIST/namvars/ cn_zoisoatl, cn_zoisoglo, cn_zoisopac, cn_zoisoinp, cn_zoisoind, cn_zoisoinp0
```
Comment (pmathiot) [inline src/modcdfnames.f90:194]: Well spotted

File: src/cdfmean.f90 [hunk 3]
```diff
@@ -93,7 +101,8 @@ PROGRAM cdfmean
   narg = iargc()
   IF ( narg == 0 ) THEN
      PRINT *,' usage : cdfmean  IN-file IN-var T|U|V|F|W [imin imax jmin jmax kmin kmax]'
-     PRINT *,'       ... [-full] [-var] [-zeromean] '
+     PRINT *,'       ... [-full] [-var] [-zeromean] [-out_suff OUT-file-suffix]'
+     PRINT *,'       ... [-basins n_basins basin_names..]'
```
Comment (pmathiot) [inline src/cdfmean.f90]: Do you mean prefix instead of suffix ? If I understand right, it is to add the path and runid (for example) before the ouput file name.

Comment (dcalve) [reply src/cdfmean.f90]: Yes I meant prefix, whoops!

File: src/cdfmean.f90 [hunk 4]
```diff
@@ -115,10 +124,15 @@ PROGRAM cdfmean
      PRINT *,'                   partial steps.'
      PRINT *,'       [ -var ]  : also compute the spatial variance of cdfvar '
      PRINT *,'       [ -zeromean ] : create a file with cdfvar having a zero spatial mean.'
+     PRINT *,'       [ -out_suff OUT-file-suffix ] : specify SUFFIX (+ directory) of the output files'
```
Comment (pmathiot) [inline src/cdfmean.f90]: I didn't really catch it before reading the code. I suggest something like "specify directory + prefix to the output name (OUT-file-suffix+cfout)"

**Review by pmathiot — COMMENTED** [review]
I agree for all the change. I would like to make some test on cdfmean and cdfmoc to be sure it is working on whatever is the time format (FLOAT or DOUBLE) as well as for some other tools. If you have a repository ready for the test, could you send me the location ? If not, don't worry, I will find something to test with me eORCA025 run.

File: src/cdfmean.f90 [hunk 5]
```diff
@@ -115,10 +124,15 @@ PROGRAM cdfmean
...
+     PRINT *,'       [ -basins n_basins basin_names.. ] : specify basins in ',TRIM(cn_fbasins),' to be used'
```
Comment (pmathiot) [inline src/cdfmean.f90]: It could be good to find a way to pickup all the basin in the file by default (ie without having to specify the number of variable and the variable name. Maybe something like a choice between your option [ -basins n_basins basin_names.. ] or [-basins_default] with no extra arguments means we pick the NEMO basin file with all the default variable names. It is a suggestion if you have time, not a requirement.

File: src/cdfmean.f90 [hunk 6]
```diff
@@ -115,10 +124,15 @@ PROGRAM cdfmean
...
      PRINT *,'      '
      PRINT *,'     REQUIRED FILES :'
      PRINT *,'       Files ', TRIM(cn_fhgr),', ', TRIM(cn_fzgr),', ', TRIM(cn_fmsk)
      PRINT *,'      '
+     PRINT *,'     OPTIONAL FILES :'
+     PRINT *,'       File ',TRIM(cn_fbasins),'; if not available only the global domain is calculated for'
+     PRINT *,'      '
```
Comment (pmathiot) [inline src/cdfmean.f90]: I suggest to change the comment by "File ',TRIM(cn_fbasins),' if -basins or -basins_default activated" because, if cn_bassin is missing and option -basins activates the tools stop (l176) and nothing is done.

File: src/cdfmean.f90 (hunk 5 above)
Comment (dcalve) [reply src/cdfmean.f90]: Does the NEMO basin file have an accepted convention in this respect? I'm not aware of an 'official' tool to produce such a file, so are "tmaskatl", "tmaskind" etc accepted standard names in the community? Are they used in NEMO?

I ask also because it may be an idea in the future to extend freely-defined basin mask specification to replace hard-coded basin mask names in some tools (e.g. `cdfmoc`). The mask itself can vary so "tmaskatl" is not in itself a standard, even if the name is.

File: src/cdfmean.f90 [hunk 7]
```diff
@@ -146,7 +160,36 @@ PROGRAM cdfmean
...
+        READ(cldum,*) nbasin
+        ALLOCATE( cbasins(nbasin) )
+
+        DO jbasin = 1, nbasin
+           IF ( ijarg > narg ) THEN
+               PRINT *, '  ERROR: n_basins must match the number of basin names provided'
+               STOP
+           ENDIF
+           CALL getarg (ijarg, cbasins(jbasin)) ; ijarg = ijarg + 1
+        END DO
```
Comment (pmathiot) [inline src/cdfmean.f90:196]: I think we should set an exit status each time we call a STOP statement (most of the cdftools do not). So I suggest to change STOP by STOP 99 .

File: src/cdfmean.f90 (hunk 6 above)
Comment (dcalve) [reply src/cdfmean.f90]: Good point

File: src/cdfmean.f90 (hunk 7 above)
Comment (dcalve) [reply src/cdfmean.f90:196]: Ok

File: src/cdfmean.f90 (hunk 5 above)
Comment (pmathiot) [reply src/cdfmean.f90]: In the commit 7692fcd of modcdfname, I move the subasin variable name and file name to the NEMO one (like this no need to rename this file and the variable). To compute quantities over a specific domain (lets say Caspian sea for example) you can build a 3d mask of the Caspian sea with cdfmkmask and run your tool with this file. I agree for eORCA12, the most convenient could be to multiply the 2d mask instead of the 3d mask. We have to be clever enough to not kill the default behaviour (ie if subbasin there, use the default variable name).

**Comment by dcalve** [comment]
Note to myself to add 'time_instant' to the list of time variables in `copyatt`:

https://github.com/met-office-ocean/CDFTOOLS/blob/3c3b651fa39c1738ba1b64a49acc886d99a625d3/src/cdfio.F90#L223-L229

File: src/cdfio.F90 [hunk 8]
```diff
@@ -1251,7 +1254,12 @@ FUNCTION  getvar (cdfile,cdvar,klev,kpi,kpj,kimin,kjmin, ktime, ldiom)
     IF (PRESENT(kimin) ) THEN
        imin=kimin
 
-       ipiglo=getdim(cdfile, cn_x, ldexact=.true.)
+       ipiglo=getdim(cdfile, cn_x, ldexact=.true., kstatus=istatus)
+       ! Else try mesh dimension name
+       IF ( istatus /= 0 ) THEN
+          ipiglo=getdim(cdfile, 'x', ldexact=.true., kstatus=istatus)
+       ENDIF
+
```
Comment (pmathiot) [inline src/cdfio.F90]: Why this change is needed ? Was it because of the x_grid_T in the files some times ago ?

Comment (dcalve) [reply src/cdfio.F90]: This change is needed for when `getvar` is used for mesh mask variables; the dimension names are still hard coded for this file.

**Comment by dcalve** [comment]
Noticed a bug here where `ldexact` is needed:

https://github.com/met-office-ocean/CDFTOOLS/blob/3c3b651fa39c1738ba1b64a49acc886d99a625d3/src/cdfmean.f90#L221

In fact I think `ldexact` should be the default behaviour for `getdim` since in most cases the dimension name is read from the namelist. 

Raised #3 for this since it will apply more widely than the functions in this pull request.

**Comment by dcalve** [comment]
Updated branch with review comment changes (209c851):

**Various**
* Add temporary non-zero stop codes

**cdfio**
* Add `time_instant` to list of dimensions in `copyatt`
* Changed default behaviour of `getdim` to `ldexact = true` (#3)
  * Will require a further review of all tools to see where `ldexact = false` is still required
* Fix `kstatus` implementation for `ldexact = true` in `getdim`

**cdfmean**
* Correct file prefix description and variable name
* Correct `cn_fbasins` description

*I haven't yet included a default option for `-basins` but will do so at some point*

**Comment by dcalve** [comment]
Updated branch with further changes (203c2322ccaa7066e9496706ecc7d953f4533585, 79c7c077669b3d37bacd52300acb124c44666d5d).

* `cdfmean` and `cdfmoc` have now been tested with ORCA2 data
* #3 can probably be closed as I can't see any case where a partial match would be needed
  * I have left `ldexact` as a keyword for `getdim` as it may be useful

**Comment by pmathiot** [comment]
I relooked all my comments, your last update and I did some test wit cdfmean. After my commit, we only need to find a good solution for the dimension issue.

**Comment by dcalve** [comment]
Added a first attempt at the dimension name matching issue (ab8637911713ca6097c65402ef74c52e9f13a4f1, see #3 )

**Comment by dcalve** [comment]
The results of `cdfmean` and `cdfmoc` have not changed with the exception of `-zeromean` output for `cdfmean`, which outputs zeroes in place of the fill_value now. This was unintentional, but I don't think it's a big issue as this is closer to the behaviour of the rest of the cdftools. The `_FillValue` attribute is what needs to be changed but is not really essential.

I haven't tested the `-rapid` or `-decomp` options of `cdfmoc` as there's no reason for the `cdfmoc` results to have changed at all.

I will go ahead and merge these changes if you are happy @pmathiot.

**Comment by pmathiot** [comment]
I agree your changes, you can merge it back to the master.
